module/zfs/ldi_iokit.cpp

```diff
@@ -1878,4 +1878,48 @@ handle_is_solidstate_iokit(struct ldi_handle *lhp, int *isssd)
 	return (0);
 }
 
+int
+handle_unmap_iokit(struct ldi_handle *lhp,
+    dkioc_free_list_t *dkm)
+{
+	int error = 0;
+
+	if (!lhp || !dkm) {
+		return (EINVAL);
+	}
+
+	if (dkm->dfl_num_exts == 0) return (0);
+
+	/* Validate IOMedia */
+	if (!OSDynamicCast(IOMedia, LH_MEDIA(lhp))) {
+		dprintf("%s invalid IOKit handle\n", __func__);
+		return (ENODEV);
+	}
+
+	LH_MEDIA(lhp)->retain();
+
+	/* We need to convert illumos' dkioc_free_list_t to dk_unmap_t */
+	IOStorageExtent *extents;
+	extents = IONew(IOStorageExtent, dkm->dfl_num_exts);
+	for (int i = 0; i < dkm->dfl_num_exts; i++) {
+		extents[i].byteStart = dkm->dfl_exts[i].dfle_start;
+		extents[i].byteCount = dkm->dfl_exts[i].dfle_length;
+	}
+	/* dkm->dfl_flags vs IOStorageUnmapOptions
+	 * #define DF_WAIT_SYNC 0x00000001 / * Wait for full write-out of free. * /
+	 * IOStorageUnmapOptions is only 0
+	 */
+
```

Comment:
Using command supplied by rottegift `fs_usage -f filesys -w | egrep -i 'unmap|trim'` shows that the `ldi_vnode.c` ioctl works, but the `ldi_iokit` call to unmap does nothing.

Reply:
As discussed on IRC, the error return should be checked as
...
if ((error = <assignment>) != kIOReturnSuccess) {
   <debug print error return>
   <handle error>
}

As opposed to checking if the error == 0 and ignoring other results (kIOReturnSuccess == 0)

And if this returns failure, zfs won’t retry trim/unmap on this vdev until it is reopened.

Reply:
Ok - fixed error check and changed returncode so it will give up trims to the device. Also know now that `unmap` calls fails due to `IOSCSIBlockCommandsDevice::IsUnmapAllowed`. Which could be due to VMware disks - might need testing on real hardware.